### PR TITLE
Add RecordFunctionGuard

### PR DESCRIPTION
**Pull request by ilia-cher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36291 Make profiler thread local
* **#36215 Add RecordFunctionGuard**

Summary:
Make it possible to disable observers, e.g. to avoid
infinite recursion if an observer uses an operator

Test Plan:
USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py develop install
./build/bin/test_jit

Differential Revision: [D20912676](https://our.internmc.facebook.com/intern/diff/D20912676)

### CODE REVIEW

File: torch/csrc/autograd/record_function.cpp [hunk 1]
```diff
@@ -218,7 +218,7 @@ void _runBeforeCallbacks(RecordFunction* rf, const std::string& funcName) {
 }
 
 RecordFunction::RecordFunction(RecordScope scope) : scope_(scope) {
-  if (manager().hasCallbacks()) {
+  if (manager().hasCallbacks() && at::_tls_is_record_function_enabled()) {
```
Comment (rohan-varma) [inline torch/csrc/autograd/record_function.cpp:221]: By default, if there is no guard, this will always be true right?

**Comment by rohan-varma** [comment]
Is there an example where an infinite recursion may happen?

**Comment by ilia-cher** [comment]
@rohan-varma yes, I believe it can happen e.g. in our NaN checker that works in observer - e.g. when you call torch::isnan it uses aten ops inside that ,in turn, call observers and so on; so we need to be careful not to call observers from observers code

File: torch/csrc/autograd/record_function.cpp (hunk 1 above)
Comment (ilia-cher) [reply torch/csrc/autograd/record_function.cpp:221]: correct

**Review by dzhulgakov — APPROVED** [review]

**Pull request closed without merging**